### DISCUSSION TITLE
Should we remove our upper version boundary on Django?

### DISCUSSION THREAD

**Discussion by laymonage** [wagtail/wagtail#12574] [Development]
## Background: pre-Wagtail 6.0

Historically, we only allowed Django versions that we officially supported at the time the Wagtail feature release came out. We received some input that the upper bound makes it difficult for people who'd like to use Wagtail with the latest version of Django. For example:

- Django 4.2 was released in April 2023 during the lifecycle of Wagtail 4.1 LTS and Wagtail 4.2 (latest).
  - Support was not added until Wagtail 5.0 (May 2023)
  - Wagtail users had to wait a month until they can use Django 4.2. They had to upgrade Wagtail first – and this was also a major release, which comes with more backwards-incompatible changes than usual.
  - This is despite Django 4.2's feature freeze (`stable/4.2.x` forked) happening 3 days before Wagtail 4.2's feature freeze (Jan 17 vs Jan 20). As long as the tests against Django's `stable/4.2.x` were passing at the time, we could've released Wagtail 4.2 with preliminary support for Django 4.2. This is because Django will not be adding any more new features to 4.2, and any regressions will be treated as release blockers.
  - If Wagtail didn't have the upper bound, users could upgrade Django a month early (when it came out), and handle upgrading to Wagtail 5.0 separately after.
- Django 5.0 was released in December 2023 during the lifecycle of Wagtail 4.2 LTS, Wagtail 5.1 (security support), and Wagtail 5.2 LTS (latest).
  - Support was not added until Wagtail 5.2.2 (2 days after Django 5.0 release) and Wagtail 6.0.
  - Not too bad this time: if you wanted to use the latest Django, you should use the latest Wagtail as well. Two days after the Django release is a pretty quick turnaround.
  - For Wagtail 5.1 and older, it's fair to _not_ support Django 5.0. The feature freeze (Django 5.0a1) was on Sep 18, and at this time Wagtail 5.1 was the latest and already more than 1.5 months in its cycle.
  - However, when Wagtail 5.2rc1 (feature freeze) was released on Oct 19, Django 5.0 was already well in feature freeze for a while. In fact, Django 5.0b1 was released just 4 days later on Oct 23.
  - Again – as long as the tests against Django's `stable/5.0.x` were passing at the time (which they **very** likely were), we could've released Wagtail 5.2<strong>.0</strong> with preliminary support for Django 5.0 simply by not having the upper bound and stating the fact that it's preliminary. This is instead of having to wait for 5.2.2.

So last year (Wagtail 6.0), [we relaxed our Django upper bound](https://github.com/wagtail/wagtail/commit/868e155c74f0b46ad6afc8461214405b03e4c7d6) to the next "major" (A.0) release.

## Wagtail 6.0 and now

Relaxing the upper bound helped quite a bit, but there are still some rough edges:

- Django 5.1 was released in August 2024 during the lifecycle of Wagtail 5.2 LTS, Wagtail 6.1 (security support), and Wagtail 6.2 (latest).
  - You could install Django 5.1 on Wagtail 6.2 and 6.1. Even also 6.0, though it was already out of support.
  - "Formal" support was not added until 6.3, and not to 6.2 at all. This is despite Django's 5.1 was already well into the beta phase (Jun 26) when Wagtail 6.2rc1 was out (Jul 19). In fact, Django 5.1rc1 was released just five days later on Jul 24. Even more, Django 5.1 **final** was released on Aug 7, **just 6 days** after Wagtail 6.2 final.
  - We added Django 5.0 support to Wagtail 5.2 in 5.2.2, why not do the same for 6.2 in 6.2.1? See the discussion in #12215.
  - Still, formalising the support sounds like it could've been an easy win for us, even if done in a patch release. Some folks like the "guarantee" of formal support when they upgrade their projects. This is even more so when Wagtail 6.2 was still the latest feature version of Wagtail.

So, recently there were questions about whether we should just drop our upper bound altogether.

## To drop or not to drop

One of the main reasons to drop the upper bound is that Django doesn't completely follow semantic versioning.

- Features deprecated in `A.0` and `A.1` are removed in `B.0`.
- Features deprecated in `A.2` (LTS) are removed in **`B.1`**. **This is the important bit.** If things can still break in non-`.0` release, is there a point in using a `<A.0` upper bound?

One of the main reasons _not_ to drop it is that some tutorials out there might do something like:

```sh
pip install wagtail>=4.1,<4.2
wagtail start mysite
```

If we didn't have the upper bound, this would install Django 5.1 and **would result in a broken project out-of-the-box**. There are two specific changes that @thibaudcolas (thank you) has confirmed break Wagtail in that scenario, where Wagtail 4.1 LTS or 5.2 LTS are used with Django 5.1:

- Removal of `STATICFILES_STORAGE` (for `wagtail start` users, as it’s in our project template)
- Removal of `Meta.index_together` (for all the ways to set up Wagtail, it comes from django-taggit)

Both of the above features were deprecated in Django 4.2 and removed in Django 5.1.

When we added support for Django 5.0 to Wagtail 5.2.2 in #11297, if we had dropped the upper boundary, or even just adopted our current "next major version" strategy (i.e. `Django<6.0`), the generated project will be broken if you do `pip install wagtail>=5.2,<5.3 && wagtail start mysite` after Django 5.1 was released.

To be fair, the breakage would only happen if you don't run `pip install -r requirements.txt`. Our project template has that file with only the officially supported Django versions allowed, and our official tutorial does tell you to run the command.

> But I thought we officially supported Django 4.2 and 5.0 in Wagtail 5.2, how come it uses features deprecated in Django 4.2?

Well, about that...

## We test against Django's `main`... right?

When questions about removing the upper bound come up, it's very easy to say:

> Well, we test against Django's `main`, so as long as the tests pass, we should be fine.

which is understandably the natural response.

**But do we, actually?**

What does it mean to test against Django's `main` branch? Well, before we get to that...

### What does it mean to **support** a Django version?

When we formalise the support for a Django version, what we do is roughly:

1. Update our CI setup to ensure we run our tests against the final releases on PyPI, as well as the `stable/A.B.x` branch in `django/django`.
4. Update our `setup.py` to allow that version.
3. Update our project template's `requirements.txt` to use that version.
2. Update our documentation and release note to "formalise" the support.

Unfortunately, it seems we don't have a good documentation for this process. In particular, we don't seem to actively check our project template against any deprecations.

For example, despite adding support for Django 4.2 in Wagtail 5.0, we did not remove the aforementioned deprecated `STATICFILES_STORAGE` until #11196, which was merged in Wagtail 6.0. That means over the course of development for Wagtail 5.0, 5.1, and 5.2, nothing in our CI triggered an error for the deprecation.

How do you define "support"? If we define it as "all tests in our test suite pass", that's OK. But if we extend the definition to "it should be possible to run a Wagtail project with that Django version without triggering any deprecation warnings" – we haven't really been following that definition. This applies to Python compatibility, too.

Case in point: I found that one of our migrations would trigger a deprecation warning on Django 5.1 due to the use of `CheckConstraint.check`: https://github.com/wagtail/wagtail/pull/12566#discussion_r1842188990

This is despite us claiming Django 5.1 support – again, depends on how we define "support".

### Our testing setup isn't complete

So, saying "tests pass against Django's `main`" may not be enough for our current test setup.

I don't think there's any automated test for the project template at all? Correct me if I'm wrong, the only point where we'd test the project template as part of a deterministic process (and not just someone happened to notice like I did above) would be when we manually test our release just before we publish them, as per https://github.com/wagtail/wagtail/wiki/Creating-a-new-Wagtail-release#package-build. Worse, in some cases the deprecation won't turn into an error, e.g. for `STATICFILES_STORAGE` it would just be completely ignored by Django in 5.1+.

I've opened https://github.com/wagtail/wagtail/pull/12566 to hopefully help catch these issues early.

Maybe we should also make it so that Django and Python's deprecation warnings raise an error when running the test suite. Not sure how feasible that is though, especially when current Wagtail is supporting a wide range of Django/Python versions.

If not, our test setup can still use some cleanup in general – right now there are deprecation/runtime warnings for our own stuff, so the output is a bit noisy. Some of them are due to deprecations that happen for code that run at startup, so using `with self.assertWarnsMessage` may not be possible. In that case, we might want to look into filtering those warnings so our test output is clean (thus easier to spot warnings we actually care about).

## So, can we drop the upper bound?

Let's say for our definition of "support", we're fine with deprecation warnings, as long as the features still work.

If our test suite passes against Django's `main`, would that mean the only blocker to removing the upper bound is the project template? If so, can we make it so that our `start` command generates a correct project, regardless of the installed Django version?

Worth noting that we support multiple Django versions – **some of which may not contain the deprecation (and its replacement) introduced in the later versions we also support**. For example, if we always immediately update our project template so as to not raise any deprecation warnings for the latest supported Django version, this means we'd also replace `STATICFILES_STORAGE` in #10310. This will not work for Django 3.2, a version that was still supported for 5.2.

So, for example, doing the following:

```
pip install Django>=3.2,<4.0  # To simulate having Django 3.2 preinstalled in your environment for whatever reason
pip install wagtail>=5.2,<5.3
wagtail start mysite
```

If we had updated the project template to use `STORAGES["staticfiles"]` instead, it would result in a "broken" project too. Well, not entirely broken – it would run, but misconfigured. And again, this would only happen if you don't run `pip install -r requirements.txt`, as otherwise you'd get Django 4.2. Essentially the same problem with running `wagtail start` with a Django version newer than expected by the template, but this time with a lower version.

**In any case, it looks like running `pip install -r requirements.txt` on a new Wagtail project is required.** Otherwise, we'd have to clutter the templates for all the possible Django versions we support. 

Django's `startproject` command (which we use for our `start` command) does give you `django_version` in the render context, so technically you can make the rendered project differ based on the installed Django version. However, this might be another can of worms, as we'd need to make sure the Django version in `requirements.txt` also matches. Otherwise if you install the requirements, you'll end up with an outdated project template with a newer Django version and having to upgrade the Django code yourself.

Which brings us to... if we expect users to `pip install -r requirements.txt`, I _think_ it should be safe to completely remove Wagtail's own upper bound on Django (but still keep it in the `requirements.txt` of the project template).

Or, maybe we could make Wagtail's `start` command smarter: if the installed Django version is newer than the one expected by the template, bail out and tell the user to install the correct version first.

Once that's done, it _should_ be safe to remove the upper bound... I think?

## Moving forward

From @thibaudcolas:

> In any case, my proposal to improve all of this would be to update our upper bound to make sure it always allows one upcoming Django version. That way the power users can always try the latest Wagtail release with the next yet-to-be-released Django, and for people who aren’t power users, it’ll still be safe.

I think this proposal makes sense, but still need clarification for the practice. By the "next" version, does this mean:
- whichever is on Django `main` at the time of the initial final release of Wagtail's latest (`A.B.0` final), or
- whichever is on Django `main` at any point during the lifetime of Wagtail's latest release (i.e. the lifetime of its `stable/a.b.x`), or
- something else, e.g. only for Django's upcoming release that are past "feature freeze"

Whichever it is, we should place the necessary safeguards to avoid any potential confusion for end users.

### Actions

- Get #12566 (or a variation of it) merged.
- Implement a better process for keeping our project template up to date. For example:
  - It currently still uses `os.path` instead of `pathlib`.
  - The `home` app doesn't define `default_auto_field`, and the settings also don't have `DEFAULT_AUTO_FIELD`. This means if you add a new model in the `home` app, you'll get `home.MyModel: (models.W042) Auto-created primary key used when not defining a primary key type, by default 'django.db.models.AutoField'.`
- Add better safeguards for our `start` command?
- Define our stance on the Django upper version boundary more clearly instead of just "the next A.0 version".
- Suggestions?

### What the future looks like

For reference, I've made a timeline view of the Django/Wagtail releases so it's easier to see the overlap between the two: https://github.com/orgs/wagtail/projects/64

(Note that exact dates may be inaccurate, the month/year should be correct though.)

With our current Django<6.0 upper bound, interesting dates:
- Sept 2025: Django 6.0a1 released.
  - Wagtail 6.3 is still the latest LTS, Wagtail 6.5* still in support, Wagtail 6.6* is the latest version.
- Dec 2025: Django 6.0 released.
  - Wagtail 6.3 LTS and 6.6* still in support, Wagtail 6.7* LTS is the latest LTS and also the latest version.

For Wagtail LTS specifically, it's going to be interesting:
- During the development of Wagtail 6.7* LTS, Django's `main` branch will shift from 6.0 to 6.1. This means Django 6.0 enters feature freeze, and can be considered "safe" for Wagtail (no more breaking changes).
- Django's 6.0 will be released after Wagtail 6.7 is released.
- Say we add official Django 6.0 support in a 6.7 patch release.
- In Wagtail 6.7's lifetime, we'll see Django's `main` branch shift yet again from 6.1 to 6.2. This will happen during the development of Wagtail 6.10*.
- Wagtail 6.10 will ship around the same time as Django 6.1, and we'll support it officially whether it's in the initial release or patch release.
- For people who use Wagtail LTS and latest Django, they can't update to the latest Django because latest Wagtail LTS is still 6.7 that only supports Django 6.0. Unless, we already bumped the upper boundary to `Django<7.0` (or removed it).
- Wagtail 6.7 wasn't actively tested against Django 6.1 during its development, as it was still actively changing even after Wagtail 6.7's release, so things could still break.
  - If we want to support Django 6.1, we need to make sure any breaking changes we detect on Wagtail's `main` get backported to `stable/6.7.x`.

*Provisional version number

**Comment by thibaudcolas** [comment]
Thank you @laymonage! @alxbridge, you oversee version upgrades across tens of sites – could we get your input on this?

This strikes me as something where it’d be great to have data to inform our decision. I have a feeling the "latest Wagtail, latest Django" crowd is pretty small, so it’d be more important to consider how it’s like for people who use a Wagtail version that’s 2-3 releases behind and still would want to be using the latest Django version.

- We could look at PyPI downloads data across versions?
- We could try a survey, like the [2023 Wagtail Deployment Survey](https://wagtail.org/blog/2023-wagtail-deployment-survey/)?
- Or just a poll on the [Wagtail Slack](https://github.com/wagtail/wagtail/wiki/Slack)?
- Not for now but future – could look into some form of opt-in or opt-out telemetry
- I could run my secret Django + Wagtail version detection script for live sites, so we see what versions they tend to run?

---

> For people who use Wagtail LTS and latest Django

Worth referencing [Wagtail release schedule](https://github.com/wagtail/wagtail/wiki/Release-schedule) and [compatible Django versions](https://docs.wagtail.org/en/stable/releases/upgrading.html#compatible-django-python-versions) here.

I’d love to hear from those people, as it seems like something that could be a common use case, but supporting it feels challenging at the point where the latest Django is being developed after our LTS release date. With no upper bound, for our users this would mean installing the latest Django and finding out about compatibility issues the hard way. For our maintainers, it’d mean backporting compatibility fixes to a release that’s meant to have very strong stability goals. Doesn’t feel impossible but certainly at odds.

> Update our upper bound to make sure it always allows one upcoming Django version

I meant it as going from an upper bound of "next 'major' (A.0)" to "next 'major' or next + 1 release, whichever is greater". Practically speaking, right now we’re at `<6.0` so there is nothing to do. Once Django 5.2 ships, we would need to update our upper bound to `<6.1` to allow Django 6.0 installations. Once Django 6.0 ships, update to `<7.0`.

This strikes me as a good solution to facilitate the work of people who want to be on the "latest Wagtail, latest Django", while not compromising on the guarantees we offer for those who are quite a few versions behind that.

> pip install -r requirements.txt

We can expect people to run it when they get started, but I think it’s still important for us to make it work anyway when they don’t. It’s unlikely but `wagtail start` still could also fail due to compatibility issues before it gets to generate that `requirements.txt`. And for power users who want to be on a version of Django that we haven’t yet tested Wagtail with, we can always suggest they make it work with [package installers supporting version resolution overrides](https://discuss.python.org/t/proposal-overrides-for-installers/23666). For example, [uv’s dependency overrides](https://docs.astral.sh/uv/concepts/resolution/#dependency-overrides).

**Reply by gasman** [reply]
Thanks @laymonage for the really thorough assessment of the current state of play! Thanks especially for pointing out the issue with Django x.1 releases not necessarily being fully backwards compatible - I gave my thumbs-up to the current "allow anything in the current major version range" on the understanding that they were. As far as I can tell that's not going to bite us any time soon, but it's definitely worth checking through the implications.

I think it might be useful to set out some expectations and "red lines" as I see them:

* When working with past Wagtail releases (either cloning an old project or pip installing an old version and running wagtail start), it's fine to require them to run pip install -r and fail otherwise. When they do run pip install -r, it should work fully with no adjustments.
* The above applies to any Wagtail release prior to the latest, even if it's still under active LTS or security support.
* When working with the latest Wagtail release, we should _aim_ to be forgiving of users failing to run pip install -r, in much the same way that we tolerate incorrect Site records up to a point. This might result in them having a "provisionally supported" version of Django - i.e. one that's permitted by Wagtail's dependency but not the project template. Ideally this should work completely with no adjustments, but it would be acceptable for it to raise deprecation warnings.
* On the day of a new Django release, it should be possible to run it against the then-current Wagtail release by updating requirements.txt, pip installing and _possibly_ fixing up to address deprecations (although if we've implemented the above point correctly that shouldn't be necessary, since we want to gracefully handle people being in this boat by mistake :-) )

As far as I can see, these conditions aren't too hard to satisfy, and still leave us plenty of options for how to deal with the Django dependency. At the most cautious end, we can pin to current minor version plus the one currently in feature freeze - as long as we deal with the new version early enough in Django's release cycle that the final is a full 3 months away, the Wagtail release will come out in time. And at the other end of the scale, I think we can ditch the upper bound - we don't have to worry about the set of "provisionally supported" Django versions stretching into the far future, because as long as Wagtail continues to release more frequently than Django, the latest Wagtail release is only ever on the hook for supporting one future Django release at most.

For the project template, we need to pin to one minor Django range - which should be the latest formally-supported one - and we just need to make sure that we stay on top of any deprecations applying to that release, so that the plus-one version continues to work (albeit possibly with deprecation warnings about version plus-two, but we'll deal with them when we formalise support for plus-one).

Please do sanity-check the above, but if that all checks out, I vote for dropping the upper bound!

**Reply by gasman** [reply]
(Sorry, that should have been a top level reply. Not used to the GitHub mobile app!)

**Comment by thibaudcolas** [comment]
Thank you @gasman! 

> Might result in them having a "provisionally supported" version of Django - i.e. one that's permitted by Wagtail's dependency but not the project template. Ideally this should work completely with no adjustments, but it would be acceptable for it to raise deprecation warnings

I think here you meant _Django_ raising deprecation warnings, but on a related note – is that scenario something we could provide a notice to users about? Something to the tune of "Heads’up, you’re running Django vx.y, which we haven’t compatibility-tested with Wagtail vx.y". I could see this being useful _if_ there are ways to do this that are visible enough, without being too obtrusive. Options I can think of are things like:

1. A possible PyPI package mechanism to flag this based on dependency resolution rules.
2. A Python-level runtime "warning" similar to the ones for deprecation but of a lower level of importance (notice / debug / info)
3. A Django-level mechanism like system checks
4. Possibly something like our [Upgrade experience for CMS users #12147](https://github.com/wagtail/wagtail/discussions/12147) in-CMS messages?

**Reply by gasman** [reply]
> I think here you meant _Django_ raising deprecation warnings

Yep, I meant Django raising deprecation warnings as a result of the Wagtail project template not doing things in the way that's expected for that Django version.

I like the sound of the incompatible-versions warning - although, are you thinking of that being raised for the "intended" provisionally-supported Django version (the one that's in beta at the time of the Wagtail release) or ones further into the future (which will only arise if they're running an older Wagtail release)? I could see the former possibly getting a bit noisy in the scenario where the Django release has just come out and will only become formally supported in the upcoming Wagtail release. And in the latter case (where we really don't expect people to be legitimately using that combination of Wagtail and Django), I wonder if there's any benefit to being clever with warnings, rather than just excluding that combination through the dependency rules. So that might be an argument in favour of a narrower dependency range...

**Reply by thibaudcolas** [reply]
I was thinking more for future releases (Django `main` at the point of the given Wagtail release). Yeah this feels more clever than warranted.

**Comment by thibaudcolas** [comment]
@laymonage (or others) – do you know how changing the upper bound would affect the experience of people upgrading to a new version? I’m wondering specifically about those types of commands that would look at version constraints to decide what to upgrade to:

```bash
pip install --update wagtail
poetry update wagtail
```

I’m not sure if Wagtail users rely on those commands without a specific target version.

**Reply by laymonage** [reply]
If I understand correctly, this command will update to the latest Wagtail version, and will try to update your installed Django version if and only if the lower bound of Django defined by Wagtail and its dependencies is higher than the currently installed version. Otherwise, it would leave your Django installation as-is.

In the case where a Django upgrade is required, I'm not sure about the default strategy for choosing which Django version will be installed, but I would assume it would be the same as if you also include `Django` in that command. In which case, it would try to resolve the latest Django version that satisfies all the boundaries defined by the project – by looking at the upper boundaries of Django defined by your dependencies (including Wagtail), or the one set in your `pyproject.toml` if you define one.

So, if we drop the upper bound, this could result in installing the latest Django version, even if it's not officially supported by the latest Wagtail yet. With the upper bound, it will only try to install the latest version that's still within the boundary (i.e. the latest supported version).

I'm not 100% sure in my answer – the exact algorithm used for dependency resolution may vary on each provider (the package manager):

- https://pip.pypa.io/en/stable/topics/dependency-resolution/
- https://pip.pypa.io/en/stable/topics/more-dependency-resolution/

**Comment by thibaudcolas** [comment]
@laymonage @gasman did we decide to proceed with a change or no in the end?

For Wagtail 6.4 released in early Feb, I believe we’re in agreement we don’t need to make any changes for compatibility with Django 5.2 as we’re already `< 6.0`. I assume we’re keeping this quiet until the Django 5.2 final release in April, and we’ll formally advertise support for Django 5.2 in Wagtail’s 6.5 May release?

Beyond that point – Django 6.0 is scheduled for release within the support period of Wagtail 6.6 and 6.7 LTS. I assume we’ll definitely want Wagtail 6.7 LTS to be compatible with Django 6.0.

- For Wagtail 6.6 – should we be looking at removing our upper version boundary or at least switching it to `< 6.1`?
- Or do that even earlier, for Wagtail 6.5? 🤔

**Comment by thibaudcolas** [comment]
Related:

- #13013 (restrict the upper bound, or backport a compatibility fix, or lockfile)
- #13012 (extend or remove the upper bound, see above)

**Comment by thibaudcolas** [comment]
Follow-up discussion about other dependencies: [Remove more upper bounds on Wagtail dependencies #13543](https://github.com/wagtail/wagtail/discussions/13543)
